src/AutoRest.CSharp.V3/Generation/Writers/JsonCodeWriterExtensions.cs

```diff
@@ -308,45 +309,73 @@ private static void CollectProperties(Dictionary<ObjectTypeProperty, CodeWriterD
+                        writer.Line($"foreach (var {collectionItemVariable:D} in {element}.EnumerateArray())");
+                        using (writer.Scope())
+                        {
+                            writer.DeserializeValue(
+                                array.ValueSerialization,
+                                w => w.Append($"{collectionItemVariable}"),
+                                (w, returnValue) => writer.Append($"{arrayVariable}.Add({returnValue});"));
+                        }
+
+                        valueCallback(writer, w => w.Append(arrayVariable));
+                    });
                     return;
                 case JsonDictionarySerialization dictionary:
-                    var dictionaryVariable = new CodeWriterDeclaration("dictionary");
-                    writer.Line($"{dictionary.Type} {dictionaryVariable:D} = new {dictionary.Type}();");
-
-                    var dictionaryItemVariable = new CodeWriterDeclaration("property");
-                    writer.Line($"foreach (var {dictionaryItemVariable:D} in {element}.EnumerateObject())");
-                    using (writer.Scope())
+                    isNullable ??= true;
+                    WriteNullableScope(() =>
                     {
-                        DeserializeValue(writer,
-                            dictionary.ValueSerialization,
-                            w => w.Append($"{dictionaryItemVariable}.Value"),
-                            (w, returnValue) => writer.Append($"{dictionaryVariable}.Add({dictionaryItemVariable}.Name, {returnValue});"));
-                    }
+                        var dictionaryVariable = new CodeWriterDeclaration("dictionary");
+                        writer.Line($"{dictionary.Type} {dictionaryVariable:D} = new {dictionary.Type}();");
 
-                    valueCallback(writer, w => w.Append(dictionaryVariable));
+                        var dictionaryItemVariable = new CodeWriterDeclaration("property");
+                        writer.Line($"foreach (var {dictionaryItemVariable:D} in {element}.EnumerateObject())");
+                        using (writer.Scope())
+                        {
+                            writer.DeserializeValue(
+                                dictionary.ValueSerialization,
+                                w => w.Append($"{dictionaryItemVariable}.Value"),
+                                (w, returnValue) => writer.Append($"{dictionaryVariable}.Add({dictionaryItemVariable}.Name, {returnValue});"));
+                        }
 
+                        valueCallback(writer, w => w.Append(dictionaryVariable));
+                    });
                     return;
                 case JsonValueSerialization valueSerialization:
-                    valueCallback(writer, w => w.DeserializeValue(valueSerialization, element));
+                    isNullable ??= valueSerialization.Type.IsNullable || !valueSerialization.Type.IsValueType;
```

Comment:
This allows all types that are nullable (meaning non-value types) to have the logic for handling JSON null values.